Fix bug of computer precision in tsearch

Jean-Romain opened the pull request:
Hi David,

This PR solve some bugs. I spent the entire day on this issue, it drove me crazy ! It was so complex to figure out what was the problem. So I must explain what I did and why I did it.

Changes in `QuadTree.cpp` correspond to nothing serious. The bug occurs when we search for a single point and probably in other cases but I spend to much time on the second bug and I don't really remember what was the problem.

The second bug, corrected in `Rtsearch.cpp` is basically the same problem than reported with `delaunayn` in #12 and #11. It is a problem of computer precision. This bug was present in the former version of the function too. 

I use a lot `tsearch` to rasterize the delaunay triangulation and I found some empty pixels not interpolated in my rasters. This happens when the searched point is very close to the edge of a triangle and the coordinates of the points are relatively large. In this case the accuracy of the barycentric coordinates and the triangle area are not accurate. Depending on the order of the points in the Delaunay matrix the function `PointInTriangle` can return both true or false. For example if the points of the triangles are in the order A B C the function return true while in the order B C A in return false because the measured areas are respectively 0.1485 and 0.1483 instead of 0.1484 (the true measure).

Therefore sometime the `tsearch` function fail to find the id of the triangle despite the fact that the point is in the triangle. It can occurs more or less randomly since the order of the point in the delaunay matrix depends on how the triangulation started. Indeed, removing one point from the original data can lead to the same triangulation but with some triangle described in another order. This minor change is sufficient to get a different result.

The solution was actually yours. I shifted the inputs to the origin (0,0). A rescaling can be done as well. In other words the `tsearch` function needs the same options than `delaunayn`. Here the `QbB` option.  

The implementation is a little bit tricky and need explanations. I have a lot of concern regarding memory usage since the implementation using QuadTree has been done to speed-up the function when used with a lot of points. Moreover the QuadTree implementation already use much more memory than the former implementation. Therefore I was not keen to create new variables to store the shifted coordinates. So, I update the original data by reference using a pointer on the user's data. That means that the user's R environment is temporarily modified since I modify the data in place. Then, I shift back the original data. This is transparent for the user.

If you're not conformable with that you can simply make a copy.

I think the shift is safe like the `QbB`option... That means *I hope it's safe*

JR

Jean-Romain commented:
Forgot my change in `QuadTree.cpp` it is stupid. I will make an update of the PR

Jean-Romain commented:
In addition I reduced the computer precision of the barycentric coordinates and the triangle area for the same reason. This modification lead to an error during the tests done in test_that. Indded the new barycentric coordinates found in `tsearch` are not exactly the same than in `tsearchn` at 10e-8 or 10e-9 since the new one are `float` and the original one `double`

Jean-Romain commented:
Well, actually it not enough yet. It remains some errors due to computer precision. For example I found a triangle for which the point was very very close to the edge and the sum of the barycentric coordinates was `s+t = 0.364500015974` and the area was `A = 0.364499986172` and therefore `s+t <= A` is false. But this point should be detected in the triangle. In my own code I tweaked the code like that: `(int)((s + t)*1000) <= (int)(A*1000)` but it is not good at all.

davidcsterratt commented:
Hi Jean-Romain, thank-you for all your work on this PR. I'm pretty busy over the next 3 days (and probably the 2 after) so I'm not going to have time to look at this properly, but I will hope to look at it late this week or early next week.

Jean-Romain commented:
No problem. I think you shouldn't accept the PR *as is* and we should talk about it before. I'm not sure of the best solution and the current PR does not solve all the problems (see third message)

davidcsterratt commented:
Sorry for the delay getting back to you - a holiday and other work intervened!

Re the issue of mutating variables to do the shifting, I don't especially like it, but I can't see anything from my quick flick through the R extensions manual to say it's _wrong_. If we were going to use that code I would perhaps run it past the r-devel email list.

Re the general issue of preventing precision errors, are the errors specific to the Quadtree implementation, or do they affect the existing implementation too? If so, perhaps the answer is to leave the functions as they are and put on a "health warning" in the documentation. It would also be interesting to see how Octave, SciPy and Matlab would deal with the input that has precision issues.

Jean-Romain commented:
The shift method is not really good and I don't like it either. It creates a side effect, and side effects are not a good practice in R. If the triangulation fails somewhere and the code does not run til the end, then the user environment is definitively changed. You can easily change that it is not a problem.

The error is not specific to the QuadTree. It comes from the way we test if a point fall inside or outside the triangle. The QuadTree only allows for fast search of points within the rectangular bounding box of the triangle. This is why the bug occurs also in the current in the CRAN release. But not necessarily for the same point or the same triangles because or the order of the computation for example. And that is more complex to manage. My PR was just to show you what I tried but I'm not convinced it is suitable in all cases.

davidcsterratt commented:
Agreed that we should avoid the side effects. 

I've tried all the new tsearch tests in this PR, and they seem to work with the version of geometry I get from install.packages(), whereas the current github master does give me the errors. Or perhaps by the remaining bug you mean this one?

> For example I found a triangle for which the point was very very close to the edge and the sum of the barycentric coordinates was s+t = 0.364500015974 and the area was A = 0.364499986172 and therefore s+t <= A is false. But this point should be detected in the triangle.

Jean-Romain commented:
Well, it was more than 2 month ago I don't remember well. 

I made extensive tests on million of triangles from triangulation of lidar data. The tests in testthat are probably specific to *my* implementation. In my implementation these tests failed before my updates.

However I don't remember if they failed with the current CRAN version. I think no. Indeed the bug is very nasty. It depends on the way the "point in triangle" test is performed and even on the order of the points in the matrix. The previous version and mine does not compute this test exactly the same way. Both were based on the barycentric method if I rememberer well but for example maybe I use points A and B while former version used point B and C to make the first cross product. This minor change implies that the bug does not occurs for the same triangles.

How I know that the previous version failed with some triangles? Because there were few holes in my rasters exactly like in my own implementation but not at the same locations.

davidcsterratt commented:
I suspect that we are running into the limits of double precision arithmetic, and from what you say I suspect that the previous implementation of tsearch in the geometry package has the same problem. I suspect that the octave implementation has the same issue too: http://hg.savannah.gnu.org/hgweb/octave/file/c2ef0eddf6bc/scripts/geometry/tsearchn.m

It may be that using multiple floating point arithmetic would help: https://cran.r-project.org/web/packages/Rmpfr/index.html  . However, this is probably a long-term piece of work.

If we could be reasonably sure that the new C++ code is no worse than the existing code, I would be minded to release, subject to all the reverse dependency checks. What do you think?

Jean-Romain commented:
I my own case I know that my data are in meter with a precision of a millimeters so my trick is relatively reliable. In your case I agree that it is a big deal difficult to solve and we are running into limit of double precision arithmetic for sure.

I just wanted to warn you about this issue but I don't have a good solution and I'm not sure there is one. If you have an idea to solve it it's fine, if not, nobody encountered the bug yet but me. And if somebody encounter the problem now you know where does it come from.

I can show you some outputs with the previous code and the new one base on million of triangle if you want to ensure that the new code is not worse than the former one. But I have to rewrite some code to regenerate them.

davidcsterratt commented:
If you could implement this fix without modifying variables in place I'd be happy to add it to the code base. I'm thinking that perhaps we could offer the quadtree algorithm and the original, inefficient algorithm as alternatives in tsearch() in version 0.4.0, which would give more chance for user feedback. My code in the tsearchn-qhull branch implements this.

Jean-Romain commented:
Well, the patch I wrote fails on 32 bits arch (testthat failed in my own package on CRAN with win 32 bits arch on regular tests). So far I can't reproduce the bug because I don't know how to compile in 32 bits and I don't want to set up a virtual machine right now.

So it is definitively not a good idea. It comes with more problems than solutions.

Jean-Romain commented:
Hi, me again.

Everything is explained here http://totologic.blogspot.com/2014/01/accurate-point-in-triangle-test.html.  The guy explains what I found myself with a trial and error method. 

The guy suggest that the barycentric method is an intrinsically less accurate method but it is only a supposition. That may explain why my code have a tendency to returns more wrong results than the previous one which was based on another method. And actually from my tests it sounds to be true.

Following this guy's advices the method I use now in my own code is the parametric equations system + point to segment distance.

So far it solved all my problems. I don't need to shift the original data, I don't have holes in my rasters anymore, it succeeded to the tricky test of computer precision in testthat even on win32 bits arch.

davidcsterratt commented:
Well done - that looks like the right way forward. I note that the orginal tserarch algortihm (now in Rtsarch_orig.c) I took from octave (
http://hg.savannah.gnu.org/hgweb/octave/file/bd77ab816e43/libinterp/corefcn/tsearch.cc ) uses the barycentric method with an extra margin. This does indeed give a negative barycentric coordinate when applied to the above example.  It also matters which order the triangulation is in, as it will put the triangle in the first triangle encountered.
```
  x1 <- 1/10
  y1 <- 1/9
  x2 <- 100/8
  y2 <- 100/3
  ## x3 = 100/4
  ## y3 = 100/9
  ## x1p = x1
  ## y1p = y1
  ## x2p = x2
  ## y2p = y2
  ## x3p = -100/8
  ## y3p = 100/6
  
  P <- rbind(c(x1, y1),
             c(x2, y2),
             c(100/4, 100/9),
             c(-100/8, 100/6))
  ## and a single point p(x, y) lying exactly on the segment [p1, p2] :
  xi <- x1 + (3/7)*(x2 - x1)
  yi <- y1 + (3/7)*(y2 - y1)

  # Gives triangle 1 as answer
  tri1 <- rbind(1:3, c(1, 2, 4))
  tsearch(P[,1], P[,2], tri1, xi, yi)
  # Also gives triangle one - even though it was triangle 2 above
  tri2 <- rbind(c(1, 2, 4), 1:3)
  tsearch(P[,1], P[,2], tri2, xi, yi)
```
It would be great if you could add this to your pull request. It would be nice to start work on getting geometry ready for CRAN. I am dreading the reverse dependency tests...

Jean-Romain commented:
>  This does indeed give a negative barycentric coordinate when applied to the above example. 

Yes. I do not compute barycentric coordinate so I don't care but geometry does. So the question is important. And the extra margin value is important. In my own case I fixed it to 1/10 millimeters but for the geometry package it is not suitable to fix it since we don't know what kind of data will be processed and in which unit.

> It also matters which order the triangulation is in, as it will put the triangle in the first triangle encountered.

Well, maybe. My point is that if the point belongs on the edge it does not matter in which triangle it will be put. I mean, the first one or the second one, its better than no one. In my own cases it does not matter but maybe you know some examples where this question matters.

> It would be great if you could add this to your pull request.

I can if we take a decision of the implantation:
* Fixed margin or not (implies an extra parameter?)
* What about negative barycentric coordinates
* How to compute barycentric coordinates efficiently from a parametric equations system? With the barycentric method the barycentric coordinates were found at the *point in triangle* time. I can compute both methods but it does not sound clever.

davidcsterratt commented:
> My point is that if the point belongs on the edge it does not matter in which triangle it will be put. I mean, the first one or the second one, its better than no one. In my own cases it does not matter but maybe you know some examples where this question matters.

I think you're right: that probably doesn't matter, certainly not for *my* problem! I use tsearch to map points from one mesh to a transformed mesh; mapping from the edge of either triangle will give the same result.

> Fixed margin or not (implies an extra parameter?)

I prefer not to add an adjustable parameter unless we really need it. I would have thought that `DOUBLE_EPS` (or some multiple of it) would be the right number, as in the original tsearch code:
https://github.com/davidcsterratt/geometry/blob/tsearchn-qhull/pkg/src/Rtsearch_orig.c
But perhaps I am missing something?

> What about negative barycentric coordinates

Set them to zero: if according to the new algorithm they are on the boundary or just inside the triangle, then setting to zero makes sense.

> How to compute barycentric coordinates efficiently from a parametric equations system? With the barycentric method the barycentric coordinates were found at the point in triangle time. I can compute both methods but it does not sound clever.

I don't know about this one! Two easy options spring to mind:
1. Are the barycentric coordinates really so bad, once there is the extra test for points with small barycentric coordinates? If so, keep the current system.
2. Just apply `cart2bary()` on the results of the QuadTree tsearch.
```
p <- do.call(rbind,
               lapply(1:nrow(xi), function(i) {
                 cart2bary(x[tri[idx[i],],], xi[i,,drop=FALSE])
}))
``` 
Not very pleasant or efficient, I agree, but as long as the code passes tests, we can try to make a more efficient version for a point release of geometry, say 0.4.1.

Jean-Romain commented:
> Are the barycentric coordinates really so bad, once there is the extra test for points with small barycentric coordinates? If so, keep the current system.

In the light of my new knowledge I finally understood the original C code of `tsearch`. In my tests barycentric method + buffer lead to few fails. This is why the original C code has also few and rare fails (barycentric + narrow buffer `1.0e-12`). But with the parametric equations system + buffer, I did not encounter any fail. It does not mean it will work every time but at least it *sounds* more robust. 

I can strictly compute like in the original C code. And if a day or another you encounter a dummy `NA` you will now that  parametric equations system might be better.

davidcsterratt commented:
On further thought, I think the conversion from the parametric to barycentric coordinates is pretty trivial, so I suggest you do compute using the the parametric coordinates and then transform them to the barycentric ones, in the C++ code. (My suggestion of doing the conversion in R will be slow for large number of points - In fact it is a problem for me at the moment.)

To show the transformation I will use a slightly different notation to the one in the post you originally referred to. Suppose the points of the triangle are P1, P2 and P3 (notionally vectors) and P is the point in the triangle. In the parametric system:
```
P = P1 + t2*(P2 - P1) + t3(P3-P1)
```
(I have used t2 and t3 instead of t1 and t2.) Rearranging:
```
P = (1 - t2 - t3)*P1 + t2*P2 + t3*P3
```
Identifying these factors with the Barycentric coordinates (b1, b2, b3) gives:
```
b1 = 1 - t2  - t3
b2 = t2
b3 = t3
```
Hope this makes sense.
